### PR TITLE
fix: OTLP exporter missing failure metrics

### PR DESCRIPTION
**Pull request by robertlaurin**
We were not capturing failures for a common failure path, we were also under
reporting failures on the final retry.

### CODE REVIEW

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb [hunk 1]
```diff
@@ -187,6 +187,7 @@ def send_bytes(bytes, timeout:) # rubocop:disable Metrics/AbcSize, Metrics/Cyclo
             when Net::HTTPBadRequest, Net::HTTPClientError, Net::HTTPServerError
               # TODO: decode the body as a google.rpc.Status Protobuf-encoded message when https://github.com/open-telemetry/opentelemetry-collector/issues/1357 is fixed.
               response.body # Read and discard body
+              @metrics_reporter.add_to_counter('otel.otlp_exporter.failure', labels: { 'reason' => response.code })
```
Comment (robertlaurin) [inline exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:190]: Currently this failure was not being reported so comparing `otel.bsp.export.failure` against `otel.otlp_exporter.failure` was not providing proper insight into export failures.

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb [hunk 2]
```diff
@@ -237,9 +238,8 @@ def measure_request_duration
         end
 
         def backoff?(retry_after: nil, retry_count:, reason:)
-          return false if retry_count > RETRY_COUNT
-
           @metrics_reporter.add_to_counter('otel.otlp_exporter.failure', labels: { 'reason' => reason })
```
Comment (robertlaurin) [inline exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:241]: We were also not reporting the final failed attempt, thanks to @SomalianIvan for catching this.

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb (hunk 1 above)
Comment (arielvalentin) [reply exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:190]: ✨

Comment (arielvalentin) [reply exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:190]: @robertlaurin Is `reason` free text other instances? Would it be sensible to use `http.status_code` as a label instead?

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb (hunk 2 above)
Comment (arielvalentin) [reply exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:241]: Any chance to add a test for this 🙏🏼 ?

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb (hunk 1 above)
Comment (robertlaurin) [reply exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:190]: I'd consider it free text at the moment
https://github.com/open-telemetry/opentelemetry-ruby/blob/469aa89ba180b1716a6a33c12647b32f1df4fed9/exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb#L200-L214

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb [hunk 3]
```diff
@@ -187,6 +187,7 @@ def send_bytes(bytes, timeout:) # rubocop:disable Metrics/AbcSize, Metrics/Cyclo
             when Net::HTTPBadRequest, Net::HTTPClientError, Net::HTTPServerError
               # TODO: decode the body as a google.rpc.Status Protobuf-encoded message when https://github.com/open-telemetry/opentelemetry-collector/issues/1357 is fixed.
```
Comment (fbogsany) [inline exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:188]: Note that the issue referenced was fixed almost a year ago. We should be able to decode the body and potentially use information it contains in the `reason` label below.

**Review by fbogsany — APPROVED** [review]
This LGTM. It’d be nice if we could decode the google.rpc.Status body, but don’t let that hold up this fix - maybe open an issue to track that?

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb (hunk 2 above)
Comment (robertlaurin) [reply exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:241]: I remember why I didn't have a test for this originally. 

The backoff behaviour basically makes running tests really slow unless you stub backoff, which is where the metric reporting behaviour is.  
https://github.com/open-telemetry/opentelemetry-ruby/blob/469aa89ba180b1716a6a33c12647b32f1df4fed9/exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb#L239-L264

It's pretty awkward to test, if you have some ideas to simplify testing this I can circle back and add it.

File: exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb (hunk 3 above)
Comment (robertlaurin) [reply exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb:188]: Tracking here https://github.com/open-telemetry/opentelemetry-ruby/issues/927

**Pull request merged**